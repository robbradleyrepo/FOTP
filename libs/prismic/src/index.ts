export * from "./customization";
export {
  linkFragment,
  metaFragment,
  pageInfoFragment,
  personFragment,
  reviewFragment,
} from "./fragments";
export * from "./link-resolver";
export { CUSTOMIZATION } from "./queries";
export * from "./rich-text";
export * from "./slices";
export * from "./types";
