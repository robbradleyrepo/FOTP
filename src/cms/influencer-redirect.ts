import {
  hasContent,
  Link,
  linkFragment,
  linkResolver,
  metaFragment,
  RichTextBlock,
  StrictMeta,
} from "@sss/prismic";
import gql from "graphql-tag";

export interface InfluencerRedirect {
  _meta: StrictMeta;
  message: RichTextBlock[] | null;
  discountCode: string | null;
  redirectUrl: Link | null;
}

export interface InfluencerRedirectData {
  influencerRedirect: InfluencerRedirect;
}

export interface TransformedInfluencerRedirect {
  destination: string;
  discountCode: string;
  handle: string;
  message: RichTextBlock[];
}

export const influencerRedirectFragment = gql`
  fragment influencerRedirect on PInfluencerRedirect {
    ...meta
    discountCode
    redirectUrl {
      ...link
    }
    message
  }
  ${linkFragment}
  ${metaFragment}
`;

export const INFLUENCER_REDIRECT = gql`
  query INFLUENCER_REDIRECT($handle: String!) {
    influencerRedirect: pInfluencerRedirect(uid: $handle) {
      ...influencerRedirect
    }
  }
  ${influencerRedirectFragment}
`;

export const transformInfluencerRedirect = ({
  _meta: { uid },
  discountCode,
  message,
  redirectUrl,
}: InfluencerRedirect): TransformedInfluencerRedirect => {
  const errors = [];
  let redirectPath: string | null = null;

  if (!discountCode?.trim()) {
    errors.push("Missing discount code");
  }

  if (!message || !hasContent(message)) {
    errors.push("Missing message");
  }

  if (!redirectUrl) {
    errors.push("Missing redirect URL");
  } else {
    redirectPath = linkResolver(redirectUrl);

    if (!redirectPath?.startsWith("/")) {
      errors.push(
        `Unable to resolve redirect URL: expected relative path, received "${redirectPath}"`
      );
    }
  }

  if (errors.length) {
    throw new Error(
      `Invalid InfluencerData:${errors
        .map((error) => `\n  â€¢ ${error}`)
        .join("")}`
    );
  }

  // Typeguard: we've already handled this
  if (!discountCode || !message || !redirectPath) {
    throw new Error("Invalid InfluencerData: unexpected result");
  }

  const uri = new URL(redirectPath, process.env.ORIGIN);

  uri.searchParams.set("influencer", uid);
  uri.searchParams.set("utm_campaign", uid);
  uri.searchParams.set("utm_medium", "share");
  uri.searchParams.set("utm_source", "influencer");

  return {
    destination: uri.toString(),
    discountCode,
    handle: uid,
    message,
  };
};
