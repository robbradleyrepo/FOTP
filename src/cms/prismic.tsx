import { CloudinaryFetchedAnimation } from "@sss/cloudinary";
import {
  ElementRendererDict,
  Elements,
  Image,
  LinkResolver,
  RichText as RichTextBase,
  RichTextBlock,
} from "@sss/prismic";
import React, { FC } from "react";

import {
  ComponentStyleProps,
  ResponsiveImageProperties,
  s,
  StyleFn,
} from "@/common/ui/utils";

import ResponsiveImage from "../ui/base/responsive-image";
import {
  headingBravo,
  headingCharlie,
  headingDelta,
  headingEcho,
} from "../ui/base/typography";
import LazyLoading from "../ui/modules/lazy-loading";

export const prismicImageToStaticImage = ({
  dimensions: { height, width },
  url,
}: Image): StaticImageData => ({
  height,
  src: url,
  width,
});

interface PrismicImageProps extends ComponentStyleProps {
  image: Image;
  sizes: string | ResponsiveImageProperties;
  priority?: true;
}

export const PrismicImage: FC<PrismicImageProps> = ({
  image,
  priority,
  sizes,
}) => {
  if (new URL(image.url).pathname.endsWith(".gif")) {
    const width = Math.min(1840, image.dimensions.width);

    return (
      <LazyLoading>
        <CloudinaryFetchedAnimation
          _css={s({
            height: "auto",
            margin: "auto",
            maxHeight: "100vh",
            width: "auto",
          })}
          url={image.url}
          width={width}
        />
      </LazyLoading>
    );
  } else {
    return (
      <ResponsiveImage
        alt={image.alt ?? ""}
        height={image.dimensions.height}
        layout="responsive"
        priority={priority}
        sizes={sizes}
        src={image.url}
        width={image.dimensions.width}
      />
    );
  }
};

export const RichTextImage: FC<
  ComponentStyleProps & (Image | Record<string, never>)
> = ({ _css = {}, alt = null, dimensions, url }) => {
  if (!dimensions || !url) {
    return null;
  }

  return (
    <figure
      css={s(
        {
          "&:last-child": {
            marginBottom: 0,
          },
          marginBottom: "1.5em",
        },
        _css
      )}
    >
      <PrismicImage image={{ alt, dimensions, url }} sizes="100vw" />
    </figure>
  );
};

export interface RichTextProps {
  components?: Partial<ElementRendererDict>;
  render: RichTextBlock[];
}

export const RichText: FC<RichTextProps> = ({ components, render }) => (
  <RichTextBase
    components={{
      [Elements.image]: <RichTextImage />,
      [Elements.paragraph]: (
        <p
          css={s({
            "&:last-child": {
              marginBottom: 0,
            },
            marginBottom: "1.5em",
          })}
        />
      ),
      ...components,
    }}
    render={render}
  />
);

const my: StyleFn = (t) => ({
  marginBottom: [t.spacing.sm, null, t.spacing.md],
  marginTop: [t.spacing.md, null, t.spacing.lg],
});

const commonListStyles = {
  paddingLeft: "1.2em",
};

const commonListItemStyles = {
  "&:last-child": {
    marginBottom: 0,
  },
  marginBottom: "1.2em",
};

export const OpinionatedRichText: FC<RichTextProps> = ({
  components,
  render,
}) => (
  <RichTextBase
    components={{
      [Elements.heading1]: (
        <h2
          css={s(headingBravo, (t) => ({
            marginBottom: [t.spacing.lg, null, t.spacing.xl],
            marginTop: [t.spacing.lg, null, t.spacing.xl],
          }))}
        />
      ),

      [Elements.heading2]: <h3 css={s(my, headingCharlie)} />,
      [Elements.heading3]: <h4 css={s(my, headingDelta)} />,
      [Elements.heading4]: <h5 css={s(my, headingEcho)} />,
      [Elements.list]: (
        <ul css={s(my, commonListStyles, { listStyle: "disc" })} />
      ),
      [Elements.hyperlink]: (
        <LinkResolver css={s({ textDecoration: "underline" })} />
      ),
      [Elements.listItem]: <li css={s(commonListItemStyles)} />,

      [Elements.oList]: (
        <ol css={s(my, commonListStyles, { listStyle: "decimal" })} />
      ),

      [Elements.oListItem]: <li css={s(commonListItemStyles)} />,

      [Elements.paragraph]: <p css={s(my)} />,
      ...components,
    }}
    render={render}
  />
);
