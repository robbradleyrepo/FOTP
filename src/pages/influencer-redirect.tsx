import { runServerSideQuery } from "@sss/apollo";
import { GetStaticPaths } from "next";

import { makeStaticPropsGetter } from "../../pages/_app";
import {
  InfluencerRedirectData,
  transformInfluencerRedirect,
} from "../cms/influencer-redirect";

const InfluencerRedirectPage = () => null;

export default InfluencerRedirectPage;

export const getStaticPaths: GetStaticPaths = async () => {
  return {
    fallback: "blocking",
    paths: [],
  };
};

export const getStaticProps = makeStaticPropsGetter(
  async ({ params }, { apolloClient }) => {
    // Typeguard: we won't hit this if the route is configured correctly
    if (typeof params?.handle !== "string") {
      return { notFound: true };
    }

    const { INFLUENCER_REDIRECT } = await import("../cms/influencer-redirect");

    const { data } = await runServerSideQuery<InfluencerRedirectData>(
      apolloClient,
      {
        query: INFLUENCER_REDIRECT,
        variables: { handle: params.handle },
      }
    );

    if (!data?.influencerRedirect) {
      return {
        notFound: true,
      };
    }

    const { destination } = transformInfluencerRedirect(
      data.influencerRedirect
    );

    return {
      redirect: {
        destination,
        permanent: false,
      },
      revalidate: 60,
    };
  }
);
