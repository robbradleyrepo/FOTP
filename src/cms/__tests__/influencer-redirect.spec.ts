import { Link, LinkType } from "@sss/prismic";

import { transformInfluencerRedirect } from "../influencer-redirect";

const invalidLinks: Link[] = [
  {
    _linkType: LinkType.document,
    _meta: {
      id: "Vr9GLaDyFKgTzB2t",
      lang: "en-us",
      tags: [],
      type: "influencer_redirect",
      uid: "test",
    },
  },
  {
    _linkType: LinkType.web,
    url: "https://us.fotp.com/account",
  },
];

describe("transformInfluencerRedirect", () => {
  it("should transform the provided `InfluencerRedirect` into `TransformedInfluencerRedirect`", () => {
    expect(
      transformInfluencerRedirect({
        _meta: {
          id: "NdHx3eubDE0YdR7I",
          lang: "en-us",
          tags: [],
          type: "influencer_redirect",
          uid: "test",
        },
        discountCode: "TEST10",
        message: [
          {
            spans: [],
            text:
              "Fans of end-to-end testing will be disappointed when they don’t get an extra 10% off game-changing dog supplements with code TEST10",
            type: "paragraph",
          },
        ],
        redirectUrl: {
          _linkType: LinkType.document,
          _meta: {
            id: "pSVcHv00qAmR9PlF",
            lang: "en-us",
            tags: [],
            type: "product_page",
            uid: "the-one",
          },
        },
      })
    ).toMatchSnapshot();

    expect(
      transformInfluencerRedirect({
        _meta: {
          id: "NdHx3eubDE0YdR7I",
          lang: "en-us",
          tags: [],
          type: "influencer_redirect",
          uid: "test",
        },
        discountCode: "TEST10",
        message: [
          {
            spans: [],
            text:
              "Fans of end-to-end testing will be disappointed when they don’t get an extra 10% off game-changing dog supplements with code TEST10",
            type: "paragraph",
          },
        ],
        redirectUrl: {
          _linkType: LinkType.web,
          url: "https://fotp.com/products/the-one?foo=bar",
        },
      })
    ).toMatchSnapshot();
  });

  it("should throw an error if any data is not set", () => {
    expect(() =>
      transformInfluencerRedirect({
        _meta: {
          id: "Hzfgl1fOn9JqJm2r",
          lang: "en-us",
          tags: [],
          type: "influencer_redirect",
          uid: "test",
        },
        discountCode: null,
        message: null,
        redirectUrl: null,
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should throw an error if an empty message is provided", () => {
    expect(() =>
      transformInfluencerRedirect({
        _meta: {
          id: "GIhDhiYlLTsp9wEL",
          lang: "en-us",
          tags: [],
          type: "influencer_redirect",
          uid: "test",
        },
        discountCode: "TEST10",
        message: [
          {
            spans: [],
            text: " ",
            type: "paragraph",
          },
        ],
        redirectUrl: {
          _linkType: LinkType.document,
          _meta: {
            id: "pSVcHv00qAmR9PlF",
            lang: "en-us",
            tags: [],
            type: "product_page",
            uid: "the-one",
          },
        },
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should throw an error if an invalid redirect URL is provided", () => {
    invalidLinks.forEach((redirectUrl) => {
      expect(() =>
        transformInfluencerRedirect({
          _meta: {
            id: "LjwxdECbjWw6U5x2",
            lang: "en-us",
            tags: [],
            type: "influencer_redirect",
            uid: "test",
          },
          discountCode: "TEST10",
          message: [
            {
              spans: [],
              text:
                "Fans of end-to-end testing will be disappointed when they don’t get an extra 10% off game-changing dog supplements with code TEST10",
              type: "paragraph",
            },
          ],
          redirectUrl,
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
