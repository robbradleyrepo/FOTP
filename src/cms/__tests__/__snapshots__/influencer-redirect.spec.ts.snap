// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformInfluencerRedirect should throw an error if an empty message is provided 1`] = `
"Invalid InfluencerData:
  • Missing message"
`;

exports[`transformInfluencerRedirect should throw an error if an invalid redirect URL is provided 1`] = `"Unable to resolve document with type \\"influencer_redirect\\" and ID \\"Vr9GLaDyFKgTzB2t\\""`;

exports[`transformInfluencerRedirect should throw an error if an invalid redirect URL is provided 2`] = `
"Invalid InfluencerData:
  • Unable to resolve redirect URL: expected relative path, received \\"https://us.fotp.com/account\\""
`;

exports[`transformInfluencerRedirect should throw an error if any data is not set 1`] = `
"Invalid InfluencerData:
  • Missing discount code
  • Missing message
  • Missing redirect URL"
`;

exports[`transformInfluencerRedirect should transform the provided \`InfluencerRedirect\` into \`TransformedInfluencerRedirect\` 1`] = `
Object {
  "destination": "https://fotp.test/products/the-one?influencer=test&utm_campaign=test&utm_medium=share&utm_source=influencer",
  "discountCode": "TEST10",
  "handle": "test",
  "message": Array [
    Object {
      "spans": Array [],
      "text": "Fans of end-to-end testing will be disappointed when they don’t get an extra 10% off game-changing dog supplements with code TEST10",
      "type": "paragraph",
    },
  ],
}
`;

exports[`transformInfluencerRedirect should transform the provided \`InfluencerRedirect\` into \`TransformedInfluencerRedirect\` 2`] = `
Object {
  "destination": "https://fotp.test/products/the-one?foo=bar&influencer=test&utm_campaign=test&utm_medium=share&utm_source=influencer",
  "discountCode": "TEST10",
  "handle": "test",
  "message": Array [
    Object {
      "spans": Array [],
      "text": "Fans of end-to-end testing will be disappointed when they don’t get an extra 10% off game-changing dog supplements with code TEST10",
      "type": "paragraph",
    },
  ],
}
`;
